sljit_src/sljitNativeARM_64.c

```diff
@@ -2517,10 +2536,62 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_atomic_store(struct sljit_compiler
 	sljit_s32 temp_reg)
 {
 	sljit_ins ins;
+	sljit_s32 mem_flags;
+	sljit_s32 src = src_reg;
+	sljit_ins cmp = 0;
 
 	CHECK_ERROR();
 	CHECK(check_sljit_emit_atomic_store(compiler, op, src_reg, mem_reg, temp_reg));
 
+#ifdef __ARM_FEATURE_ATOMICS
+	if (op & SLJIT_SET_ATOMIC_STORED)
+		cmp = SUBS ^ W_OP;
+
+	switch (GET_OPCODE(op)) {
+	case SLJIT_MOV32:
+	case SLJIT_MOV_U32:
+		ins = CAS ^ (1 << 30);
+		mem_flags = INT_SIZE;
+		break;
+	case SLJIT_MOV_U8:
+		ins = CASB;
+		mem_flags = BYTE_SIZE;
+		break;
+	case SLJIT_MOV_U16:
+		ins = CASH;
+		mem_flags = HALF_SIZE;
+		break;
+	default:
+		ins = CAS;
+		mem_flags = WORD_SIZE;
+		if (cmp)
+			cmp ^= W_OP;
+		break;
+	}
+
+	FAIL_IF(push_inst(compiler, ins | RM(temp_reg) | RN(mem_reg) | RD(src)));
+	if (!cmp)
+		return SLJIT_SUCCESS;
+
+	switch(mem_flags) {
+	case HALF_SIZE:
+		FAIL_IF(push_inst(compiler, (ANDI ^ W_OP) | RN(src) | RD(TMP_REG1) | (0x1e << 9)));
+		src = TMP_REG1;
+		break;
+	case BYTE_SIZE:
+		FAIL_IF(push_inst(compiler, (ANDI ^ W_OP) | RN(src) | RD(TMP_REG1) | (0xe << 9)));
+		src = TMP_REG1;
+		break;
+	}
+
+	FAIL_IF(emit_op_mem(compiler, mem_flags, TMP_REG2, SLJIT_MEM1(mem_reg), 0, TMP_REG2));
```

Comment:
this works in practice, but it would be ideal if replaced with `push_inst()` specially to avoid having to rely on `emit_op_mem()` not obfuscating the value from TMP_REG2.

alternatively `temp_reg` could be used instead